[WIP] feature(application): Introduces \Elgg\Application object and moved most of the booting process into it.

Srokap opened the pull request:
THIS IS WORK IN PROGRESS. DO NOT MERGE.

I'm adding it as a working preview of changes, but still needs some more features.

It's single step toward #7721 that aims to remove dependency on start.php file location. Only path the user should need is path to it's project autoloader file and do not care about Elgg internals.

TODO:
- [ ] better \Elgg\Config management
- [ ] allowing to pass inline configuration as alternative to loading it from file
- [ ] allow to customize settings file path, decide on most convenient api.
- [x] prevent repeated triggering of booting steps on single app instance
- [ ] deprecate $CONFIG? doubt we'll be able to throw it away at this point. Need to have a look and think.
- [ ] update handlers code
- [ ] update icondirect code

Srokap committed:
Merge branch '1.x'

ewinslow committed:
Merge pull request #7030 from Elgg/1.x

Merge 1.x up to master

ewinslow committed:
Merge pull request #7181 from Elgg/1.x

Merging 1.x into master

Srokap committed:
Merge branch '1.x'

Srokap committed:
chore(docs): Bumped docs version and badges to indicate master version

ewinslow committed:
Merge pull request #7690 from Srokap/master

Merge branch '1.x' into master

mrclay committed:
Merge branch '1.x'

ewinslow committed:
Merge pull request #7828 from mrclay/master_1x

Merging 1.x into master

mrclay committed:
fix(breadcrumbs): hide the last breadcrumb if it’s not a link

Fixes #6419

mrclay committed:
Merge pull request #7830 from mrclay/6419_breadcrumb

fix(breadcrumbs): hide the last breadcrumb if it’s not a link

jdalsem committed:
chore(messages): remove unused references to msg metadata

fixes #5948

mrclay committed:
Merge pull request #7856 from jdalsem/remove-msg-metadata-3

chore(messages): remove unused references to msg metadata

jdalsem committed:
fix(site_notifications): correctly position site_notifications menu

fixes #6809

jdalsem committed:
chore(engine): drop support for elgg_version in manifests

Also removes the Elgg 1.7 manifest related tests and classes.

fixes #4323

jdalsem committed:
fix(views): do not drop elgg-button-submit class when using input/submit

Also the input/button and input/submit views can now handle
$vars['class'] to be an array.

Srokap committed:
feature(application): Introduces \Elgg\Application object and moved most of the booting process into it.

Deprecated inclusion of start.php for booting the core.

Refs #7721

mrclay commented on install.php:
```diff
@@ -12,7 +12,7 @@
 }
 
-require_once __DIR__ . "/vendor/autoload.php";
+require_once "/vendor/autoload.php";
```
I don't understand this change

Srokap replied:
I'll have to double check that for correctness, but have problems with elgg used as dependency, where autoloader is in vendor dir on root level, not relativley to `/vendor/elgg/elgg`. Try running Elgg using https://github.com/Srokap/elgg_as_dep, works reasonably well already, but things like icons don't work yet, due to mod dir being still kept in `/vendor/elgg/elgg`.

Srokap replied:
P.S. What I was affraid of: http://stackoverflow.com/questions/25438287/how-to-configure-a-composer-package-to-be-globally-installed#answer-25438654

Will have to handle dependency case manually in Elgg files.

ewinslow replied:
Safe to assume elgg is installed at vendor/elgg/elgg?

mrclay commented:
First impression is great.

ewinslow commented on engine/classes/Elgg/Application.php:
```diff
@@ -0,0 +1,293 @@
+ *
+ * @access private
+ *
+ * @package    Elgg.Core
```
No more package or sub package annotations needed.

ewinslow commented on engine/classes/Elgg/Application.php:
```diff
@@ -0,0 +1,293 @@
+
+
+		// No settings means a fresh install
+		if (!is_file("{$this->engineDir}/settings.php")) {
```
I have a nice OO filesystem API in core now that you might consider using...

Srokap replied:
FYI it's just the existing code moved around. It certainly could use improvements, will have a look at it.

ewinslow replied:
Sgtm

ewinslow commented:
Very cool. Two issues:

 - Can we do Elgg\Application::start($wwwRoot)?
 - The rewrite rules getting duplicated in php is a little brutal... Any way we can simplify that?

ewinslow commented:
Man I would really love to see this happen for 2.0...

jdalsem committed:
Merge pull request #7904 from jdalsem/drop-elgg-version-in-manifest

chore(engine): drop support for elgg_version in manifests

mrclay commented on engine/classes/Elgg/Application.php:
```diff
@@ -0,0 +1,293 @@
+	/**
+	 * @var string
+	 */
+	private $wwwRoot;
```
"wwwRoot" sounds like a URL. I'd probably prefer "elggDir".

ewinslow replied:
rootDir? elggDir sounds like vendor/elgg/elgg which I don't think is what we want either

ewinslow replied:
Ah, I'm mixing up PRs, but still. We should distinguish between the root of elgg and the root of the app since those may not be the same going forward...

mrclay commented on engine/classes/Elgg/AutoloadManager.php:
```diff
@@ -193,5 +193,9 @@ public function getLoader() {
 		$this->storage = $storage;
 	}
+
+	public function __destruct() {
```
This may work, but I've had several bad experiences relying on PHP destructors, particularly calling out to other objects (that may have already been destroyed).

mrclay replied:
This case is probably OK.

mrclay commented on engine/classes/Elgg/CacheHandler.php:
```diff
@@ -249,7 +249,8 @@ protected function renderView($view, $viewtype) {
 	 */
 	protected function loadEngine() {
-		require_once dirname(dirname(dirname(__FILE__))) . "/start.php";
+		$app = new \Elgg\Application();
```
My gut reaction is that the Application should be passed into CacheHandler

ewinslow replied:
+1. Yay DI

mrclay commented on engine/classes/Elgg/Application.php:
```diff
@@ -0,0 +1,293 @@
+	 *
+	 * You need to explicitly point to index.php in order for router to work properly:
+	 *
+	 * <code>php -S localhost:8888 index.php</code>
```
Is there an easy way to set the web root when using this? I assume currently the external resource links are broken?

Srokap replied:
Using `php -S localhost:8888 index.php` is totally fine regardless. That's not a new thing, just moved in single place. I use it for testing Elgg for some time now and it handles everything properly.

ewinslow commented:
Would love to keep this moving along. I really think we ought to shoot for this in 2.0.

ewinslow committed:
chore(i18n): Move 2.0 to its own Transifex project

We need to do this since there will be backwards incompatible changes
to language strings in 2.0 most likely and there is no way to keep
Transifex from syncing those changes back to 1.x without creating a
new project.

ewinslow committed:
Merge branch '1.x' into 1.x-to-master

Conflicts:
	docs/guides/plugins/dependencies.rst

ewinslow committed:
Merge pull request #8022 from ewinslow/i18n-new-transifex-project

chore(i18n): Move 2.0 to its own Transifex project

Commit:
Merge pull request #8026 from ewinslow/1.x-to-master

Merge 1.x to master

physcocode commented:
You nailed i man +1

ewinslow commented:
Would it help move this along if you just replaced the body of start.php with a call to `Elgg\Application::start()` so you don't have to do so many more of these refactors? Also helps with BC

Srokap commented:
Ignore me - testing.

beck24 committed:
Merge branch '1.x-merge' into master-merge

Conflicts:
	docs/guides/upgrading.rst

mrclay committed:
fix(database): Uncallable callback arguments now throw exceptions

Fixes #6937

jdalsem committed:
Merge pull request #8132 from beck24/master-merge

chore(merge): Merge 1.x up to Master

jdalsem committed:
Merge pull request #8137 from jdalsem/notifications-menu

fix(site_notifications): correctly position site_notifications menu

mrclay committed:
chore(upgrades): Removes pre-1.11 upgrade scripts, actions, and views

Also removes 3 deprecated functions that use ext/mysql resources directly

Fixes #5703

mrclay committed:
Merge pull request #8155 from mrclay/db_callbacks_6937

fix(database): Uncallable callback arguments now throw exceptions

jdalsem committed:
Merge pull request #8135 from jdalsem/button-class-issues

fix(views): do not drop elgg-button-submit class when using input/submit

Srokap committed:
Moved autoloader path choosing logic into new autoloader.php file

mrclay committed:
Merge pull request #8164 from mrclay/remove_mysql_20

chore(upgrades): Removes pre-1.11 upgrade scripts, actions, and views

juho-jaakkola committed:
Merge branch '1.x'

mrclay committed:
Merge pull request #8179 from juho-jaakkola/merge_1x_to_master

Merge 1.x to master

mrclay committed:
chore(views): Entity/annotation views no longer rely on ->view

Elgg no longer allows $entity->view metadata nor $annotation->view to
influence the view used for rendering.

Fixes #8017

mrclay committed:
Merge pull request #8185 from mrclay/view_md_8017

chore(views): Entity/annotation views no longer rely on ->view

mrclay committed:
Moves to 2.0, allows injecting settings file location, minor refactoring

mrclay committed:
Removes more unwanted start.php usage.
Allows Elgg\Config or setting location injection in constructor.
Doesn’t allow multiple loads/boots of core.
Use Application in CacheHandler.
Require Application when setting up DIC, inject config service.
Fix notice checking CONFIG in views.
Export handler works in CLI server.

mrclay committed:
Cleaner Application constructor

mrclay committed:
Constructor @param format

mrclay committed:
Clarified @param in _elgg_services()

mrclay committed:
Cleanup use of CONFIG in service provider

mrclay committed:
No notices checking isHttps()

mrclay committed:
Inject Config object so the application doesn’t try to load settings.php

mrclay committed:
Fix @param in service provider

mrclay committed:
Set default site_guid in installer

mrclay committed:
Use singleton instance to get phpunits working

mrclay committed:
Remove accidental code

mrclay committed:
Only set $START_MICROTIME once

mrclay committed:
Get rid of global plugin-loading functions and some unneeded internal state

mrclay committed:
Moves SP into Application, and moves settings loading into Config

mrclay committed:
Cleanup now that SP is an Application property

juho-jaakkola commented:
Mostly replaced by #8247 so closing.
